### PR TITLE
Feat: add basic EFI library

### PR DESCRIPTION
**Pull request by Abb1x**
I added a basic EFI library for the work-in-progress bootloader.

I am using my own printf-like thing because for a certain reason, fmt wouldn't want to work with utf16.
I also implemented my own `strncpy` since yours is broken (add a '\0' at the end of the string).

### CODE REVIEW

**Review by sleepy-monax — CHANGES_REQUESTED** [review]

File: sources/libs/efi/base.h [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#define EFI_SUCCESS 0
+
+#define UNUSED(x) (void)x
+
+typedef unsigned short char16;
+typedef unsigned long long EFIStatus;
+typedef void *EFIEvent;
+typedef void *EFIHandle;
+
+typedef uint8_t u8;
```
Comment (Supercip971) [inline sources/libs/efi/base.h:16]: Why using u16 u32… here ? And if you want special type for uefi maybe add a namespace

File: sources/libs/efi/misc.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#ifndef __CLIB_H__
+#define __CLIB_H__
```
Comment (Supercip971) [inline sources/libs/efi/misc.h:2]: Pragma once

**Review by Supercip971 — CHANGES_REQUESTED** [review]

File: sources/libs/efi/base.h (hunk 1 above)
Comment (Abb1x) [reply sources/libs/efi/base.h:16]: it was easier to type when I was declaring the efi functions. :sweat:

**Comment by gek169** [comment]
strncpy is not supposed to "place a \0 at the end of the string"
Strncpy is supposed to fill the unused portion of the string with \0
If the entirety of 'len' is occupied by the string, there will be no null terminator.

As such, it is normal to write your code such that strncpy receives `len-1` rather than `len` if len is the lengtrh of your buffer, and for you to write `\0` into the last byte yourself.

**Comment by gek169** [comment]
this is a bad fix.

**Comment by gek169** [comment]
http://port70.net/~nsz/c/c89/c89-draft.html#4.11.2.4

**Comment by Abb1x** [comment]
> strncpy is not supposed to "place a \0 at the end of the string"
> Strncpy is supposed to fill the unused portion of the string with \0
> If the entirety of 'len' is occupied by the string, there will be no null terminator.
> 
> As such, it is normal to write your code such that strncpy receives `len-1` rather than `len` if len is the length of your buffer, and for you to write `\0` into the last byte yourself.

yeah this was removed after

**Pull request closed without merging**